src/data/set/finite.lean

```diff
@@ -20,529 +47,733 @@ variables {α : Type u} {β : Type v} {ι : Sort w} {γ : Type x}
+set.fintype_bUnion' s f
 
-@[simp] theorem finite.insert (a : α) {s : set α} : finite s → finite (insert a s)
-| ⟨h⟩ := ⟨@set.fintype_insert _ (classical.dec_eq α) _ _ h⟩
+instance fintype_empty : fintype (∅ : set α) := fintype.of_finset ∅ $ by simp
 
-lemma to_finset_insert [decidable_eq α] {a : α} {s : set α} (hs : finite s) :
-  (hs.insert a).to_finset = insert a hs.to_finset :=
-finset.ext $ by simp
+instance fintype_singleton (a : α) : fintype ({a} : set α) := fintype.of_finset {a} $ by simp
 
-@[simp] lemma insert_to_finset [decidable_eq α] {a : α} {s : set α} [fintype s] :
-  (insert a s).to_finset = insert a s.to_finset :=
-by simp [finset.ext_iff, mem_insert_iff]
+instance fintype_pure : ∀ a : α, fintype (pure a : set α) :=
+set.fintype_singleton
 
-@[elab_as_eliminator]
-theorem finite.induction_on {C : set α → Prop} {s : set α} (h : finite s)
-  (H0 : C ∅) (H1 : ∀ {a s}, a ∉ s → finite s → C s → C (insert a s)) : C s :=
-let ⟨t⟩ := h in by exactI
-match s.to_finset, @mem_to_finset _ s _ with
-| ⟨l, nd⟩, al := begin
-    change ∀ a, a ∈ l ↔ a ∈ s at al,
-    clear _let_match _match t h, revert s nd al,
-    refine multiset.induction_on l _ (λ a l IH, _); intros s nd al,
-    { rw show s = ∅, from eq_empty_iff_forall_not_mem.2 (by simpa using al),
-      exact H0 },
-    { rw ← show insert a {x | x ∈ l} = s, from set.ext (by simpa using al),
-      cases multiset.nodup_cons.1 nd with m nd',
-      refine H1 _ ⟨finset.subtype.fintype ⟨l, nd'⟩⟩ (IH nd' (λ _, iff.rfl)),
-      exact m }
-  end
-end
+/-- A `fintype` structure on `insert a s` when inserting a new element. -/
+def fintype_insert_of_not_mem {a : α} (s : set α) [fintype s] (h : a ∉ s) :
+  fintype (insert a s : set α) :=
+fintype.of_finset ⟨a ::ₘ s.to_finset.1, s.to_finset.nodup.cons (by simp [h]) ⟩ $ by simp
 
-@[elab_as_eliminator]
-theorem finite.dinduction_on {C : ∀s:set α, finite s → Prop} {s : set α} (h : finite s)
-  (H0 : C ∅ finite_empty)
-  (H1 : ∀ {a s}, a ∉ s → ∀ h : finite s, C s h → C (insert a s) (h.insert a)) :
-  C s h :=
-have ∀ h : finite s, C s h,
-  from finite.induction_on h (λ h, H0) (λ a s has hs ih h, H1 has hs (ih _)),
-this h
+/-- A `fintype` structure on `insert a s` when inserting a pre-existing element. -/
+def fintype_insert_of_mem {a : α} (s : set α) [fintype s] (h : a ∈ s) :
+  fintype (insert a s : set α) :=
+fintype.of_finset s.to_finset $ by simp [h]
 
-instance fintype_singleton (a : α) : fintype ({a} : set α) :=
-unique.fintype
+instance fintype_insert (a : α) (s : set α) [decidable_eq α] [fintype s] :
+  fintype (insert a s : set α) :=
+fintype.of_finset (insert a s.to_finset) $ by simp
 
-@[simp] theorem card_singleton (a : α) :
-  fintype.card ({a} : set α) = 1 :=
-fintype.card_of_subsingleton _
+/-- Normally, `insert a s` for `finset` requires `[decidable_eq α]`, but we can use this weaker
+assumption here. -/
+instance fintype_insert' (a : α) (s : set α) [decidable $ a ∈ s] [fintype s] :
```

Comment:
Having this instance as well as the previous seems a bit odd to me, but can we reference the other?

Reply:
This is from the original `fintype_insert`, which derives decidable membership from decidable equality and chooses one of the two `fintype` instances based on that decision. The new `fintype_insert` just uses the `finset` insert, which makes it more like instances for other set operations -- having looked at the history, it looks like the reason for this is that the original `set.finite` was a custom inductive type that made no use of `finset`.  I let the original `fintype_insert` hang around in this form since it might still be useful.